File: packages/artifact/python/ferrule_artifact/artifact.py
import base64
import hashlib

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from .canonical import canonicalize


def build(manifest: object | bytes | str) -> bytes:
    """Build an artifact as canonical manifest bytes."""
    return canonicalize(manifest)


def artifact_hash(artifact: bytes) -> str:
    """Hash an artifact previously produced by build()."""
    return "sha256:" + hashlib.sha256(artifact).hexdigest()


def _digest(artifact: bytes) -> bytes:
    return hashlib.sha256(artifact).digest()


def sign(artifact: bytes, private_key_pem: bytes) -> bytes:
    key = serialization.load_pem_private_key(private_key_pem, password=None)
    if not isinstance(key, Ed25519PrivateKey):
        raise ValueError("private key is not Ed25519")
    return key.sign(_digest(artifact))


def verify(artifact: bytes, signature: bytes, public_key_pem: bytes) -> bool:
    """Verify a detached signature, returning False for all invalid input."""
    if len(signature) != 64:
        return False
    try:
        key = serialization.load_pem_public_key(public_key_pem)
        if not isinstance(key, Ed25519PublicKey):
            return False
        key.verify(signature, _digest(artifact))
    except (InvalidSignature, TypeError, ValueError):
        return False
    return True


def generate_dev_keypair() -> tuple[bytes, bytes]:
    """Generate a local development keypair; production keys need managed custody."""
    private_key = Ed25519PrivateKey.generate()
    private_pem = private_key.private_bytes(
        serialization.Encoding.PEM,
        serialization.PrivateFormat.PKCS8,
        serialization.NoEncryption(),
    )
    public_pem = private_key.public_key().public_bytes(
        serialization.Encoding.PEM,
        serialization.PublicFormat.SubjectPublicKeyInfo,
    )
    return private_pem, public_pem


def encode_signature(signature: bytes) -> str:
    return base64.b64encode(signature).decode("ascii")


def decode_signature(encoded: str) -> bytes | None:
    try:
        return base64.b64decode(encoded, validate=True)
    except ValueError:
        return None

File: packages/artifact/go/artifact.go
package artifact

import (
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"crypto/x509"
	"encoding/hex"
	"encoding/pem"
	"errors"
)

func Build(manifest []byte) ([]byte, error) {
	return Canonicalize(manifest)
}

func Hash(artifact []byte) string {
	digest := sha256.Sum256(artifact)
	return "sha256:" + hex.EncodeToString(digest[:])
}

func Sign(artifact []byte, privateKey ed25519.PrivateKey) ([]byte, error) {
	if len(privateKey) != ed25519.PrivateKeySize {
		return nil, errors.New("invalid Ed25519 private key length")
	}
	digest := sha256.Sum256(artifact)
	return ed25519.Sign(privateKey, digest[:]), nil
}

func Verify(artifact, signature []byte, publicKey ed25519.PublicKey) bool {
	if len(signature) != ed25519.SignatureSize || len(publicKey) != ed25519.PublicKeySize {
		return false
	}
	digest := sha256.Sum256(artifact)
	return ed25519.Verify(publicKey, digest[:], signature)
}

// GenerateDevKeypairPEM is only for local development. Production keys require
// managed key custody and must never be written by this helper.
func GenerateDevKeypairPEM() ([]byte, []byte, error) {
	publicKey, privateKey, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return nil, nil, err
	}
	privateDER, err := x509.MarshalPKCS8PrivateKey(privateKey)
	if err != nil {
		return nil, nil, err
	}
	publicDER, err := x509.MarshalPKIXPublicKey(publicKey)
	if err != nil {
		return nil, nil, err
	}
	return pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: privateDER}),
		pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: publicDER}), nil
}

func ParsePrivateKeyPEM(data []byte) (ed25519.PrivateKey, error) {
	block, _ := pem.Decode(data)
	if block == nil {
		return nil, errors.New("invalid private key PEM")
	}
	key, err := x509.ParsePKCS8PrivateKey(block.Bytes)
	if err != nil {
		return nil, err
	}
	privateKey, ok := key.(ed25519.PrivateKey)
	if !ok {
		return nil, errors.New("private key is not Ed25519")
	}
	return privateKey, nil
}

func ParsePublicKeyPEM(data []byte) (ed25519.PublicKey, error) {
	block, _ := pem.Decode(data)
	if block == nil {
		return nil, errors.New("invalid public key PEM")
	}
	key, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return nil, err
	}
	publicKey, ok := key.(ed25519.PublicKey)
	if !ok {
		return nil, errors.New("public key is not Ed25519")
	}
	return publicKey, nil
}
